File: arrays_and_strings/zero_matrix.py
import unittest


def nullify_row(matrix, row):
    for i in range(len(matrix[0])):
        matrix[row][i] = 0


def nullify_col(matrix, col):
    for j in range(len(matrix)):
        matrix[j][col] = 0


def zero_matrix(matrix):
    m = len(matrix)
    n = len(matrix[0])
    rows = []
    cols = []

    for i in range(m):
        for j in range(n):
            if matrix[i][j] == 0:
                rows.append(i)
                cols.append(j)

    for row in rows:
        nullify_row(matrix, row)

    for col in cols:
        nullify_col(matrix, col)

    return matrix


class Test(unittest.TestCase):
    '''Test Case'''
    data = [
        ([
            [1, 2, 3, 4, 0],
            [6, 0, 8, 9, 10],
            [11, 12, 13, 14, 15],
            [16, 0, 18, 19, 20],
            [21, 22, 23, 24, 25]
        ], [
            [0, 0, 0, 0, 0],
            [0, 0, 0, 0, 0],
            [11, 0, 13, 14, 0],
            [0, 0, 0, 0, 0],
            [21, 0, 23, 24, 0]
        ])
    ]

    def test_zero_matrix(self):
        for [test_matrix, expected] in self.data:
            actual = zero_matrix(test_matrix)
            self.assertEqual(actual, expected)


if __name__ == '__main__':
    unittest.main()

File: arrays_and_strings/zero_matrix.cpp
#include <iostream>

void nullifyRow(int ** matrix, int row, int col) {
    for (int j = 0; j < col; j++)
        matrix[row][j] = 0;
}

void nullifyCol(int ** matrix, int row, int col) {
    for (int i = 0; i < row; i++)
        matrix[i][col] = 0;
}

void nullifyMatrix(int ** matrix, int row, int col) {
    bool firstRow = false;

    for  (int i = 0; i < col; i++) {
        if (matrix[0][i] == 0) {
            firstRow = true;
            break;
        }
    }

    for (int i = 1; i < row; i++) {
        bool nullifyThisRow = false;
        for (int j = 0; j < col; j++) {
            if (matrix[i][j] == 0) {
                matrix[0][j] = 0;
                nullifyThisRow = true;
            }
        }
        if (nullifyThisRow)
            nullifyRow(matrix, i, col);
    }

    // We know which column to be nullify using information from previous step.
    for (int j = 0; j < col; j++) {
        if (matrix[0][j] == 0)
            nullifyCol(matrix, row, j);
    }

    // nullify the first row if it has zero
    if (firstRow)
        nullifyRow(matrix, 0, col);
}

void printMatrix(int ** matrix, int row, int col) {
    for (int i = 0; i < row; i++) {
        for (int j = 0; j < col; j++)
            std::cout << matrix[i][j] << " ";
        std::cout << '\n';
    }
    std::cout << '\n';
}
int main() {
    int row, col;
    std::cout << "Enter number of rows: ";
    std::cin >> row;
    std::cout << "Enter number of cols: ";
    std::cin >> col;
    int ** matrix = new int*[row];
    for (int i = 0; i < row; i++) {
        matrix[i] = new int[col];
    }
    std::cout << "Provide MxN matrix " << '\n';
    for (int i = 0; i < row; i++) {
        for (int j = 0; j < col; j++) {
            std::cin >> matrix[i][j];
        }
    }
    std::cout << "Matrix before: " << '\n';
    printMatrix(matrix, row, col);
    nullifyMatrix(matrix, row, col);
    std::cout << "Matrix after: " << '\n';
    printMatrix(matrix, row, col);
    return 0;
}
